README.md
# Amarok Software Center
Software Center for Amarok Linux OS



debian/changelog
amarok-software (0.3.0-1amarok2) stable; urgency=medium

  * New upstream version for Amarok Linux OS
  * update

 -- Amarok Linux <packages@amaroklinux.org>  Wed, 10 Aug 2022 13:47:42 -0300

amarok-software (0.3.0-1amarok1) stable; urgency=medium

  * New upstream version for Amarok Linux OS
  * Brazilian portuguese translation added

 -- Amarok Linux <packages@amaroklinux.org>  Tue, 09 Aug 2022 12:01:16 -0300

amarok-software (0.3.0-1) stable; urgency=medium

  * Rebuild for Amarok Linux OS

 -- Amarok Linux <packages@amaroklinux.org>  Fri, 05 Aug 2022 23:26:21 -0300

pardus-software (0.3.0) unstable; urgency=medium

  [ Fatih Altun ]
  * Add control for server app command package names
  * Fix myapps listbox creating bug on some old systems
  * Add authorization configuration to use repo and myapps sections
  * Add animation controls for myapps stack
  * Add pardus-software group and polkit for passwordless usage
  * Update pkla actions as full name
  * Add passwordless usage control button to settings
  * Add another apt update way to exceptions
  * Add show update button after fix sources and update info label ui
  * Update group action error label, discard 32256
  * Update pot file
  * Update Turkish translations
  * Add update-translations script
  * Add justify center to dAptUpdateButton
  * Add reset scroll position to comments

  [ yakusha ]
  * Correct question mark punctiation

  [ Fatih Altun ]
  * Update pardus-software.conf as interactive conf file
  * Rm postrm script
  * Add GLib.idle to set pardusappsdetail page
  * Add row activated action to myapps listbox
  * Add app command data control to myapps activated func
  * Update rating ui to be more useful user experience
  * Add activeapp control to rating
  * Fix multiple widget create problem
  * Update pot file
  * Update Turkish translations

 -- Fatih Altun <fatih.altun@pardus.org.tr>  Wed, 20 Jul 2022 21:01:31 +0300

pardus-software (0.2.0) unstable; urgency=medium

  * Add extra desktop name controls to controlargs function
  * Add missing lower method to search func
  * Add search in repo apps too on controlargs func
  * Fix repo apps searching controls in controlargs
  * Fix if app not found value in controlargs
  * Add control style_context on controlargs func
  * Add package name tooltip to app name
  * Add installed, required, freed size infos with required changes to pardus app detail
  * Update file size units
  * Fix typo
  * Update pot file
  * Update Turkish translations
  * Fix show available apps config control
  * Add thread for get package size and required changes info
  * Update size and required changes info after action completed
  * Add GLib_idle to size_worker_done func
  * Update sleep time as 0.25
  * Update size and required changes infos as empty before loaded
  * Add subcategory combobox to pardus apps
  * Add subcategory key control for old servers
  * Add user's all desktop apps to myapps and Update myapps ui
  * Add GB to get size method
  * Add except for broken packages while marking
  * Update to_install and to_remove controls
  * Add details page for myapps' remove action
  * Add action to myapps' open button
  * Add localization to get installed_apps
  * Add disclaimer to myapps' remove action for important package removings
  * Add remove action to myapps and Fix get myapps queue icon
  * Add control_myapps func for control myapps list after remove
  * Add updatecache to myapp details page
  * Add action to myapps' uninstall disclaimer accept button
  * Fix sensitive control of myapps_uninstall_button
  * Add status,error,cache control to control_myapps func
  * Rm unnecessary code on myapplistbox remove
  * Revert updatecache on myapp details page for better loading time
  * Add remove option to application
  * Add search ability for myapps too
  * Fix typo
  * Update myapps menu sensitive control on no connection
  * Add menuback controls for myapps page
  * Update get_installed_apps and parse_desktopfile funcs using Gio
  * Update myapps' notfound page
  * Update menu back controls
  * Fix typo and Add sleep control to required_changes
  * Fix typo
  * Update pot file
  * Update Turkish translations
  * Update pardus app detail action button as buttonbox with info button
  * Add dActionInfoButton sensitive controls
  * Add required changes popover to daction info button
  * Add key_press_event control for myapps page too
  * Update RequiredChangesPopover ui
  * Add row_activated action to queue listbox
  * Fix set size labels on worker_done func
  * Fix menuback sensitive control on pardus page click
  * Add subcatcombo clear to most and last buttons on home
  * Update pot file
  * Update Turkish translations
  * Add get important packages from server if available

 -- Fatih Altun <fatih.altun@pardus.org.tr>  Thu, 09 Jun 2022 10:30:31 +0300

pardus-software (0.1.1) unstable; urgency=medium

  [ Hugo Carvalho ]
  * Add Portuguese (pt) translation
  * Add files via upload

  [ Fatih Altun ]
  * Add read more button to pardus apps description
  * Add mo file create function as dynamically to setup.py
  * Add gettext to Build-Depends
  * Fix window overflow problem on small resolutions
  * Add completed info label to queue box
  * Update pot file
  * Update Turkish translations
  * Add rm apt lists to fix sources list func

 -- Fatih Altun <fatih.altun@pardus.org.tr>  Mon, 25 Apr 2022 16:27:40 +0300

pardus-software (0.1.0) unstable; urgency=medium

  * Fix if homepage app not in Pardus apps
  * Update copyright to 2022
  * Update get_locale function for more accurate detection
  * Add exception for handle notify errors
  * Add libgtk-3-bin to Depends

 -- Fatih Altun <fatih.altun@pardus.org.tr>  Mon, 28 Mar 2022 12:17:06 +0300

pardus-software (0.1.0~beta30) unstable; urgency=medium

  * Add controlServer func for handle Soup exceptions

 -- Fatih Altun <fatih.altun@pardus.org.tr>  Thu, 10 Mar 2022 16:44:10 +0300

pardus-software (0.1.0~beta29) unstable; urgency=medium


src/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Sep 18 14:53:00 2020

@author: fatih
"""

import gi, sys
from MainWindow import MainWindow

gi.require_version('Gtk', '3.0')
from gi.repository import Gtk, Gio, GLib


class Application(Gtk.Application):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, application_id="org.amarok.software",
                         flags=Gio.ApplicationFlags(8), **kwargs)
        self.window = None

        self.add_main_option(
            "details",
            ord("d"),
            GLib.OptionFlags(0),
            GLib.OptionArg(1),
            "Details page of application",
            None,
        )

        self.add_main_option(
            "remove",
            ord("r"),
            GLib.OptionFlags(0),
            GLib.OptionArg(1),
            "Remove page of application",
            None,
        )

    def do_activate(self):
        # We only allow a single window and raise any existing ones
        if not self.window:
            # Windows are associated with the application
            # when the last one is closed the application shuts down
            self.window = MainWindow(self)
        else:
            self.window.controlArgs()
        self.window.MainWindow.present()

    def do_command_line(self, command_line):
        options = command_line.get_options_dict()
        options = options.end().unpack()
        self.args = options
        self.activate()
        return 0


if __name__ == "__main__":
    app = Application()
    app.run(sys.argv)


src/Server.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Sep 18 14:53:00 2020

@author: fatih
"""

import requests
from pathlib import Path
import tarfile
from shutil import rmtree
from hashlib import md5

import gi, json

gi.require_version("GLib", "2.0")
gi.require_version('Soup', '2.4')
from gi.repository import GLib, Gio, Soup


class Server(object):
    def __init__(self):
        self.serverurl = ""  # This is setting from MainWindow server func
        self.serverapps = "/api/v2/apps/"
        self.servercats = "/api/v2/cats/"
        self.serverhomepage = "/api/v2/homepage"
        self.serverstatistics = "/api/v2/statistics"
        self.serversendrate = "/api/v2/rate"
        self.serversenddownload = "/api/v2/download"
        self.serversendsuggestapp = "/api/v2/suggestapp"
        self.serverparduscomments = "/api/v2/parduscomments"
        self.serverfiles = "/files/"
        self.serverappicons = "appicons"
        self.servercaticons = "categoryicons"
        self.servericonty = ".svg"
        self.serverarchive = ".tar.gz"
        self.serversettings = "/api/v2/settings"
        self.settingsfile = "serversettings.ini"

        userhome = str(Path.home())
        try:
            self.username = userhome.split("/")[-1]
        except:
            self.username = ""
        self.cachedir = userhome + "/.cache/amarok-software/"
        self.configdir = userhome + "/.config/amarok-software/"

        self.error_message = ""
        self.connection = False
        self.applist = []
        self.orgapplist = []
        self.catlist = []
        self.ediapplist = []
        self.mostdownapplist = []
        self.mostrateapplist = []
        self.lastaddedapplist = []
        self.totalstatistics = []
        self.servermd5 = []
        self.appversion = ""
        self.iconnames = ""
        self.badwords = []
        self.dailydowns = []
        self.osdowns = []
        self.appdowns = []
        self.oscolors = []
        self.appcolors = []
        self.aptuptime = 86400  # default control value is 1 day if server value is none

        self.gnomeratingserver = "https://odrs.gnome.org/1.0/reviews/api/ratings"
        self.gnomecommentserver = "https://odrs.gnome.org/1.0/reviews/api/fetch"

    def get(self, url, type):
        file = Gio.File.new_for_uri(url)
        file.load_contents_async(None, self._open_stream, type)

    def _open_stream(self, file, result, type):
        try:
            success, data, etag = file.load_contents_finish(result)
        except GLib.Error as error:
            self.error_message = error.message
            print("{} _open_stream Error: {}, {}".format(type, error.domain, error.message))
            self.ServerAppsCB(False, response=None, type=type)  # Send to MainWindow
            return False

        if success:
            self.ServerAppsCB(True, json.loads(data), type)
        else:
            print("{} is not success".format(type))
            self.ServerAppsCB(False, response=None, type=type)  # Send to MainWindow

    def getIcons(self, url, type, force_download=False, fromsettings=False):
        if not self.isExists(self.cachedir + type) or force_download:
            file = Gio.File.new_for_uri(url)
            file.load_contents_async(None, self._open_icon_stream, type, fromsettings)
        else:
            print("{} already available".format(type))
            self.ServerIconsCB(True, type, fromsettings)

    def _open_icon_stream(self, file, result, type, fromsettings):
        try:
            success, data, etag = file.load_contents_finish(result)
        except GLib.Error as error:
            print("{} _open_icon_stream Error: {}, {}".format(type, error.domain, error.message))
            self.ServerIconsCB(False, type, fromsettings)
            return False

        if success:
            if self.createDir(self.cachedir):
                with open(self.cachedir + type + self.serverarchive, "wb") as file:
                    file.write(data)
                if self.controlMD5(type):
                    if self.extractArchive(self.cachedir + type + self.serverarchive, type):
                        self.ServerIconsCB(True, type, fromsettings)
                        return True
                    else:
                        print("{} extract error".format(type))
                else:
                    print("md5 value is different (controlMD5) {} ".format(type))
        else:
            print("{} is not success".format(type))

        self.ServerIconsCB(False, type, fromsettings)

    # def getAppIcons(self, force_download=False):
    #     if not self.isExists(self.cachedir + self.serverappicons) or force_download:
    #         print("trying to downlad " + self.serverappicons)
    #         try:
    #             response = requests.get(self.serverurl + self.serverfiles + self.serverappicons + self.serverarchive)
    #         except:
    #             print(
    #                 "server error getting " + self.serverurl + self.serverfiles + self.serverappicons + self.serverarchive)
    #             return False
    #         if response.status_code == 200:
    #             if self.createDir(self.cachedir):
    #                 with open(self.cachedir + self.serverappicons + self.serverarchive, "wb") as file:
    #                     file.write(response.content)
    #                 if self.controlMD5(self.serverappicons):
    #                     if self.extractArchive(self.cachedir + self.serverappicons + self.serverarchive,
    #                                            self.serverappicons):
    #                         return True
    #                     else:
    #                         print("extract error")
    #                         return False
    #                 else:
    #                     print("md5 value is different (controlMD5)")
    #                     return False
    #             return False
    #         else:
    #             print("{} : {}".format("error getting app icons, status code", response.status_code))
    #             return False
    #     else:
    #         return True

    # def getCategoryIcons(self, force_download=False):
    #     if not self.isExists(self.cachedir + self.servercaticons) or force_download:
    #         print("trying to downlad " + self.servercaticons)
    #         try:
    #             response = requests.get(self.serverurl + self.serverfiles + self.servercaticons + self.serverarchive)
    #         except:
    #             print(
    #                 "server error getting " + self.serverurl + self.serverfiles + self.servercaticons + self.serverarchive)
    #             return False
    #         if response.status_code == 200:
    #             if self.createDir(self.cachedir):
    #                 with open(self.cachedir + self.servercaticons + self.serverarchive, "wb") as file:
    #                     file.write(response.content)
    #                 if self.controlMD5(self.servercaticons):
    #                     if self.extractArchive(self.cachedir + self.servercaticons + self.serverarchive,
    #                                            self.servercaticons):
    #                         return True
    #                     else:
    #                         print("extract error")
    #                         return False
    #                 else:
    #                     print("md5 value is different (controlMD5)")
    #                     return False
    #             return False
    #         else:
    #             print("{} : {}".format("error getting category icons, status code", response.status_code))
    #             return False
    #     else:
    #         return True

    def controlIcons(self):
        redown_app_icons = False
        redown_cat_icons = False
        if self.isExists(self.cachedir + self.serverappicons + self.serverarchive):
            localiconmd5 = md5(open(self.cachedir + self.serverappicons + self.serverarchive, "rb").read()).hexdigest()
            if self.servermd5["appicon"]:
                if localiconmd5 != self.servermd5["appicon"]:
                    print("md5 value of app icon is different so trying download new app icons from server")
                    redown_app_icons = True

        if self.isExists(self.cachedir + self.servercaticons + self.serverarchive):
            localiconmd5 = md5(open(self.cachedir + self.servercaticons + self.serverarchive, "rb").read()).hexdigest()
            if self.servermd5["caticon"]:
                if localiconmd5 != self.servermd5["caticon"]:
                    print("md5 value of cat icon is different so trying download new cat icons from server")
                    redown_cat_icons = True

        return redown_app_icons, redown_cat_icons

    def controlMD5(self, type):
        if type == self.serverappicons:
            servertag = "appicon"
        elif type == self.servercaticons:
            servertag = "caticon"
        else:
            return False

        if self.isExists(self.cachedir + type + self.serverarchive):
            localiconmd5 = md5(open(self.cachedir + type + self.serverarchive, "rb").read()).hexdigest()
            if self.servermd5[servertag]:
                if localiconmd5 == self.servermd5[servertag]:
                    return True
        return False

    # def getDefaultSettings(self):
    #     if not self.isExists(self.configdir):
    #         print("trying to get settings from server")
    #         try:
    #             response = requests.get(self.serverurl + self.serversettings)
    #         except:
    #             print("server error getting " + self.serverurl + self.serversettings)
    #             return False
    #         if response.status_code == 200:
    #             if self.createDir(self.configdir):
    #                 with open(self.configdir + self.settingsfile, "wb") as file:
    #                     file.write(response.content)
    #                     return True
    #         else:
    #             print("{} : {}".format("error getting settings, status code", response.status_code))
    #             return False
    #     else:
    #         return True

    def createDir(self, dir):
        try:
            Path(dir).mkdir(parents=True, exist_ok=True)
            return True
        except:
            print("{} : {}".format("mkdir error", self.cachedir))
            return False

    def extractArchive(self, archive, type):
        try:
            tar = tarfile.open(archive)
            if type == self.serverappicons:
                extractables = [member for member in tar.getmembers() if
                                member.name.startswith(self.serverappicons) and member.name.endswith(self.servericonty)]
            elif type == self.servercaticons:
                extractables = [member for member in tar.getmembers() if
                                member.name.startswith(self.servercaticons) and member.name.endswith(self.servericonty)]
            else:
                extractables = ""
            rmtree(self.cachedir + type, ignore_errors=True)
            try:
                icons = self.iconnames.split(",")
                for icon in icons:
                    rmtree(self.cachedir + type + "-" + icon, ignore_errors=True)
            except Exception as e:
                print("{}".format(e))
            tar.extractall(members=extractables, path=self.cachedir)
            tar.close()
            return True
        except:
            print("tarfile error")
            return False

    def isExists(self, dir):
        if Path(dir).exists():
            print(dir + " exists")
            return True
        else:
            print(dir + " not exists")
            return False

    # def getGnomeRatings(self):
    #     try:
    #         request_gnome = requests.get(self.gnomeratingserver)
    #     except Exception as e:
    #         print(e)
    #         print("Connection problem on gnome odrs / ratings")
    #         self.gnomeconnection = False
    #
    #     if self.gnomeconnection:
    #         if request_gnome.status_code == 200:
    #             return request_gnome.json()
    #         return False
    #     return False

    def deleteCache(self):
        try:
            rmtree(self.cachedir)
            return True, ""
        except Exception as e:
            print(str(e))
            return False, str(e)


src/UserSettings.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Sep 18 14:53:00 2020

@author: fatih
"""

from pathlib import Path
import configparser, distro


class UserSettings(object):
    def __init__(self):

        try:
            self.userdistroid = distro.id().lower()
        except:
            self.userdistroid = ""
        try:
            self.userdistroversion = distro.major_version().lower()
        except:
            self.userdistroversion = ""
        try:
            self.usercodename = distro.codename().lower()
            if self.usercodename == "Rolling":
                self.usercodename = self.usercodename + self.userdistroversion
        except:
            self.usercodename = ""
        try:
            self.userdistro = ", ".join(distro.linux_distribution())
        except:
            self.userdistro = ""

        userhome = str(Path.home())
        try:
            self.username = userhome.split("/")[-1]
        except:
            self.username = ""
        self.configdir = userhome + "/.config/amarok-software/"
        self.configfile = "settings.ini"
        self.config = configparser.ConfigParser(strict=False)
        self.config_usi = None
        self.config_ea = None
        self.config_saa = None
        self.config_hera = None
        self.config_icon = None
        self.config_sgc = None
        self.config_udt = None
        self.config_aptup = None
        self.config_lastaptup = None
        self.config_forceaptuptime = None

    def createDefaultConfig(self, force=False):
        self.config['DEFAULT'] = {'UseServerIcons': 'yes',
                                  'Animations': 'yes',
                                  'ShowAvailableApps': 'yes',
                                  'HideExternalRepoApps': 'yes',
                                  'IconName': 'default',
                                  'ShowGnomeComments': 'yes',
                                  'UseDarkTheme': 'no',
                                  'AutoAptUpdate': 'yes',
                                  'LastAutoAptUpdate': '0',
                                  'ForceAutoAptUpdateTime': '0'}

        if not Path.is_file(Path(self.configdir + self.configfile)) or force:
            if self.createDir(self.configdir):
                with open(self.configdir + self.configfile, "w") as cf:
                    self.config.write(cf)

    def readConfig(self):
        try:
            print("in readconfig")
            self.config.read(self.configdir + self.configfile)
            self.config_usi = self.config.getboolean('DEFAULT', 'UseServerIcons')
            self.config_ea = self.config.getboolean('DEFAULT', 'Animations')
            self.config_saa = self.config.getboolean('DEFAULT', 'ShowAvailableApps')
            self.config_hera = self.config.getboolean('DEFAULT', 'HideExternalRepoApps')
            self.config_icon = self.config.get('DEFAULT', 'IconName')
            self.config_sgc = self.config.getboolean('DEFAULT', 'ShowGnomeComments')
            self.config_udt = self.config.getboolean('DEFAULT', 'UseDarkTheme')
            self.config_aptup = self.config.getboolean('DEFAULT', 'AutoAptUpdate')
            self.config_lastaptup = self.config.getint('DEFAULT', 'LastAutoAptUpdate')
            self.config_forceaptuptime = self.config.getint('DEFAULT', 'ForceAutoAptUpdateTime')
        except Exception as e:
            print("{}".format(e))
            print("user config read error ! Trying create defaults")
            # if not read; try to create defaults
            self.config_usi = True
            self.config_ea = True
            self.config_saa = True
            self.config_hera = True
            self.config_icon = "default"
            self.config_sgc = True
            self.config_udt = False
            self.config_aptup = True
            self.config_lastaptup = 0
            self.config_forceaptuptime = 0
            try:
                self.createDefaultConfig(force=True)
            except Exception as e:
                print("self.createDefaultConfig(force=True) : {}".format(e))

    def writeConfig(self, srvicons, anims, avaiapps, extapps, iconname, gnomecom, darktheme, aptup, lastaptup, faptupt):
        self.config['DEFAULT'] = {'UseServerIcons': srvicons,
                                  'Animations': anims,
                                  'ShowAvailableApps': avaiapps,
                                  'HideExternalRepoApps': extapps,
                                  'IconName': iconname,
                                  'ShowGnomeComments': gnomecom,
                                  'UseDarkTheme': darktheme,
                                  'AutoAptUpdate': aptup,
                                  'LastAutoAptUpdate': lastaptup,
                                  'ForceAutoAptUpdateTime': faptupt}
        if self.createDir(self.configdir):
            with open(self.configdir + self.configfile, "w") as cf:
                self.config.write(cf)
                return True
        return False

    def createDir(self, dir):
        try:
            Path(dir).mkdir(parents=True, exist_ok=True)
            return True
        except:
            print("{} : {}".format("mkdir error", dir))
            return False


src/SysActions.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Sep 18 14:53:00 2020

@author: fatih
"""

import os
import subprocess
import sys
from shutil import rmtree
from pathlib import Path
import distro
from datetime import datetime
import apt
import random, string


def main():
    def update():
        try:
            cache = apt.Cache()
            cache.open()
            cache.update()
        except Exception as e:
            print(str(e))
            subupdate()

    def subupdate():
        subprocess.call(["apt", "update"],
                        env={**os.environ, 'DEBIAN_FRONTEND': 'noninteractive'})

    def fixbroken():
        subprocess.call(["apt", "install", "--fix-broken", "-yq"],
                        env={**os.environ, 'DEBIAN_FRONTEND': 'noninteractive'})

    def dpkgconfigure():
        subprocess.call(["dpkg", "--configure", "-a"],
                        env={**os.environ, 'DEBIAN_FRONTEND': 'noninteractive'})

    def aptclean():
        subprocess.call(["apt", "clean"],
                        env={**os.environ, 'DEBIAN_FRONTEND': 'noninteractive'})

    def externalrepo(key, sources, name):
        tmpkeyfilename = ''.join(random.choice(string.ascii_lowercase) for i in range(13))
        tmpkeyfile = open(os.path.join("/tmp", tmpkeyfilename), "w")
        tmpkeyfile.write(key)
        tmpkeyfile.flush()
        tmpkeyfile.close()
        tmpkey = os.path.join("/tmp", tmpkeyfilename)

        subprocess.call(["apt-key", "add", tmpkey])

        Path("/etc/apt/sources.list.d/").mkdir(parents=True, exist_ok=True)
        sdfile = open("/etc/apt/sources.list.d/" + name, "w")
        sdfile.write(sources)
        sdfile.flush()
        sdfile.close()

        if os.path.isfile(tmpkey):
            os.remove(tmpkey)

    def correctsourceslist():
        found = True
        source = ""
        codename = distro.codename().lower()
        now = datetime.now()

        if codename == "bullseye":
            source = "### The Official Amarok Package Repositories ###\n\n" \
                     "deb http://repo.amaroklinux.org/amarok bullseye main contrib non-free\n" \
                     "deb http://deb.debian.org/debian bullseye main contrib non-free\n" \
                     "# deb-src http://deb.debian.org/debian bullseye main contrib non-free\n\n" \
		     "deb http://deb.debian.org/debian bullseye-updates main contrib non-free\n" \
                     "# deb-src http://deb.debian.org/debian bullseye-updates main contrib non-free\n\n" \
                     "deb http://deb.debian.org/debian-security bullseye-security main contrib non-free\n" \
                     "# deb-src http://deb.debian.org/debian-security bullseye-security main contrib non-free\n\n" \
                     "### This section generated by Amarok Software at " + str(now) + " ###\n"
        elif codename == "testing":
            source = "### The Official Amarok Package Repositories ###\n\n" \
                     "deb http://repo.amaroklinux.org/rolling rolling main contrib non-free\n" \
                     "deb http://deb.debian.org/debian testing main contrib non-free\n" \
                     "# deb-src http://deb.debian.org/debian testing main contrib non-free\n\n" \
		     "deb http://deb.debian.org/debian testing-updates main contrib non-free\n" \
                     "# deb-src http://deb.debian.org/debian testing-updates main contrib non-free\n\n" \
                     "deb http://security.debian.org/debian-security testing-security main contrib non-free\n" \
                     "# deb-src http://security.debian.org/debian-security testing-security main contrib non-free\n\n" \
                     "### This section generated by Amarok Software at " + str(now) + " ###\n"
            else:
                found = False        
        elif codename == "bookworm":
            source = "### The Official Debian Package Repositories ###\n\n" \
		       "deb http://deb.debian.org/debian bookworm main contrib non-free\n" \
                     "# deb-src http://deb.debian.org/debian bookworm main contrib non-free\n\n" \
		       "deb http://deb.debian.org/debian bookworm-updates main contrib non-free\n" \
                     "# deb-src http://deb.debian.org/debian bookworm-updates main contrib non-free\n\n" \
                       "deb http://security.debian.org/debian-security bookworm-security main contrib non-free\n" \
                     "# deb-src http://security.debian.org/debian-security bookworm-security main contrib non-free\n\n" \
                     "### This section generated by Amarok Software at " + str(now) + " ###\n"
        else:
            found = False

        if found:
            rmtree("/etc/apt/sources.list.d", ignore_errors=True)
            rmtree("/var/lib/apt/lists/", ignore_errors=True)
            Path("/etc/apt/sources.list.d").mkdir(parents=True, exist_ok=True)
            sfile = open("/etc/apt/sources.list", "w")
            sfile.write(source)
            sfile.flush()
            sfile.close()
            aptclean()

    if len(sys.argv) > 1:
        if sys.argv[1] == "externalrepo":
            externalrepo(sys.argv[2], sys.argv[3], sys.argv[4])
            update()
        elif sys.argv[1] == "correctsourceslist":
            correctsourceslist()
        elif sys.argv[1] == "update":
            update()
        elif sys.argv[1] == "fixapt":
            correctsourceslist()
            subupdate()
            fixbroken()
            dpkgconfigure()
        else:
            print("unknown argument error")
    else:
        print("no argument passed")


if __name__ == "__main__":
    main()
